### PR TITLE
Replaced PANDAS_DATETIMEUNIT with NPY_DATETIMEUNIT

### PR DESCRIPTION
**Pull request by WillAyd**
progress towards #21852

- [X] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Working on my C-fu as well. There may be a better way to do this (ex: even adding NPY_DATETIMEUNIT to Cython/includes) but figured I'd offer this up for review

@jbrockmendel

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pxd [hunk 1]
```diff
@@ -33,3 +32,21 @@ cdef maybe_datetimelike_to_i8(object val)
 cdef int64_t tz_convert_utc_to_tzlocal(int64_t utc_val, tzinfo tz)
 
 cpdef datetime localize_pydatetime(datetime dt, object tz)
+
+cdef extern from "numpy/ndarraytypes.h":
```
Comment (WillAyd) [inline pandas/_libs/tslibs/conversion.pxd]: Not sure if there is a better way to do this declaration. Right now I copy/pasted in the three files where it is need but is there a shared place to put that for shared access across the modules?

Comment (jbrockmendel) [inline pandas/_libs/tslibs/conversion.pxd]: Since this is done in np_datetime.pxd, I'd just cimport from there rather than copying this.

**Comment by jbrockmendel** [comment]
Aside from that note on repeated imports, this looks like a step in the right direction.  The other super-easy change-over is `pandas_datetimestruct`.

File: pandas/_libs/tslibs/conversion.pxd (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pxd]: Since it is in np_datetime.pxd, it can be `cimport`ed from there.

**Comment by jbrockmendel** [comment]
I still can’t find a way to access functions in numpy’s _datetime.h.  Not even sure what is possible in C-land.  @chris-b1 are you the right person to ask about this?

**Comment by WillAyd** [comment]
@jbrockmendel not sure what you are looking for out of that but it doesn't appear that header is included in numpy's `core/include` directory along with the rest of the headers. You might have to mess around with `setup.py` if you are trying to include header files outside of that location:

https://github.com/pandas-dev/pandas/blob/5d0daa0522730a6f999ceaf328f63f03dd62d0b4/setup.py#L128

Though that most likely violates a lot of best practices

**Comment by WillAyd** [comment]
Though looking at it I don't believe `src/multiarray/_datetime.h` is included as part of your numpy installation in site-packages so may not be possible with how the code is currently structured (could be wrong)

**Comment by jbrockmendel** [comment]
> not sure what you are looking for out of that but it doesn't appear that header is included in numpy's core/include directory along with the rest of the headers

The hope was that it is somehow passed through via an `#include` in one of the files that is part of core/include.  After all, numpy itself somehow manages to call these functions

**Comment by chris-b1** [comment]
Yeah, even if it were possible I don't think we should be depending on private API of numpy - presumably why some bits (e.g. datetime string parsing) were vendored in the first place.

**Comment by jreback** [comment]
cc @charris

do you recall if these existed in numpy < 1.6 ?

**Comment by WillAyd** [comment]
AFAICT goes at least back to 1.4

**Comment by jbrockmendel** [comment]
AFAICT the code was vendored b/c at the time pandas supported older numpy versions that did not yet have `datetime64`.  Also because pandas handles timezones differently in `np_datetime_strings`.  It'd still be nice to not re-implement the large portion that is identical.

It would also be super-convenient to be able to use the numpy versions in src/ujson/python/obj2JSON.c, as that would let us de-couple the remaing src/datetime code.

**Comment by jreback** [comment]
thanks @WillAyd and @jbrockmendel for the comments. Agree that the vendored code, just wanted the rationale documented here :>

**Pull request merged**